File: 0028/solution.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        if len(needle) > len(haystack):
            return -1
        N = len(needle)
        H = len(haystack)
        MOD = 10**9 + 7
        PRIME = 131
        powers = [1]
        for i in range(1, H):
            powers.append((powers[i - 1] * 31) % MOD)
        needle_hash = 0
        for i in range(N):
            needle_hash = (needle_hash + ord(needle[i]) * powers[i]) % MOD
        haystack_hash = {-1: 0}
        for i in range(H):
            haystack_hash[i] = (
                haystack_hash[i - 1] + ord(haystack[i]) * powers[i]
            ) % MOD
        for i in range(N - 1, H):
            curr_needle_hash = (needle_hash * powers[i - (N - 1)]) % MOD
            curr_haystack_hash = (haystack_hash[i] - haystack_hash[i - N]) % MOD
            if curr_needle_hash == curr_haystack_hash:
                return i - (N - 1)
        return -1

File: 0028/solution.java
class Solution {
    public int strStr(String h, String n) {
        int H = h.length();
        int N = n.length();
        int MOD = 1_000_000_007;
        if (N > H)
            return -1;

        long[] powers = new long[H];
        final int PRIME = 131;
        powers[0] = 1;
        for (int i = 1; i < H; i++) {
            powers[i] = ((long) powers[i - 1] * PRIME) % MOD;
        }

        long[] prefixes = new long[H];
        prefixes[0] = h.charAt(0);
        for (int i = 1; i < H; i++) {
            prefixes[i] = (prefixes[i - 1] + (h.charAt(i)) * powers[i]) % MOD;
        }

        long needle_hash = n.charAt(0);
        for (int i = 1; i < N; i++) {
            needle_hash = (needle_hash + (n.charAt(i)) * powers[i]) % MOD;
        }

        for (int i = 0; i <= H - N; i++) {
            int end = i + N - 1;
            long haystack_hash = prefixes[end];
            if (i > 0) {
                haystack_hash = (haystack_hash - prefixes[i - 1] + MOD) % MOD;
            }

            if (haystack_hash == (needle_hash * powers[i]) % MOD) {
                return i;
            }
        }

        return -1;
    }
}
